README.md

```diff
@@ -0,0 +1,556 @@
+into JSON. However it means more complicated code as info needs to be
+passed between the services in http requests, more for the user to
+set-up and manage - another app config, another apache config, separate
+log files.
+
+XLoader - the job runs in a worker process, in the same app as CKAN, so
+can access the CKAN config, db and logging directly and avoids many HTTP
+calls. This simplification makes sense because the xloader job doesn't
+need to do much processing - mainly it is streaming the CSV file from
+disk into PostgreSQL.
+
+It is still entirely possible to run the XLoader worker on a separate
+server, if that is desired. The worker needs the following:
+
+- A copy of CKAN installed in the same Python virtualenv (but not
+  running).
+- A copy of the CKAN config file.
+- Access to the Redis instance that the running CKAN app uses to store
+  jobs.
+- Access to the database.
+
+You can then run it via ckan jobs worker as below.
+
+### Caveat - column types
+
+Note: With XLoader, all columns are stored in DataStore's database as
+'text' type (whereas DataPusher did some rudimentary type guessing -
+see 'Robustness' above). However once a resource is xloaded, an admin
+can use the resource's Data Dictionary tab to change these types to
+numeric or datestamp and re-load the file. When migrating from
+DataPusher to XLoader you can preserve the types of existing resources
+by using the `migrate_types` command.
+
+There is scope to add functionality for automatically guessing column
+type -offers to contribute this are welcomed.
+
+## Requirements
+
+Compatibility with core CKAN versions:
+
+  | CKAN version   | Compatibility                                                                                                  |
+  | -------------- | -------------------------------------------------------------------------------------------------------------- |
+  | 2.7            | no longer supported (last supported version: 0.12.2)                                                           |
+  | 2.8            | no longer supported (last supported version: 0.12.2)                                                           | 
+  | 2.9            | yes (Python3) (last supported version for Python 2.7: 0.12.2)), Must: `pip install "setuptools>=44.1.0,<71"`   |
+  | 2.10           | yes                                                                                                            |
+  | 2.11           | yes                                                                                                            |
+
+## Installation
+
+To install XLoader:
+
+1.  Activate your CKAN virtual environment, for example:
+
+        . /usr/lib/ckan/default/bin/activate
+
+2.  Install the ckanext-xloader Python package into your virtual
+    environment:
+
+        pip install ckanext-xloader
+
+3.  Install dependencies:
+
+        pip install -r https://raw.githubusercontent.com/ckan/ckanext-xloader/master/requirements.txt
```

Comment:
I think it's time this is cleaned up as using the latest instead of what you installed via pypi or git is not very good in the scheme of things.

For www.data.qld.gov.au, we use this bash like script to do the find of requirements to install.
see https://github.com/qld-gov-au/opswx-ckan-cookbook/blob/dde4e48ac6c50e663c364afec462224dcf7fc9a1/resources/pip_install_app.rb#L114C17-L134C66
(we may not need all of this but this is what we use to cover all plugins out in the wild atm)
```
PYTHON_MAJOR_VERSION=$(#{new_resource.virtualenv_dir}/bin/python -c "import sys; print(sys.version_info.major)")
PYTHON_REQUIREMENTS_FILE=requirements-py$PYTHON_MAJOR_VERSION.txt
if [ -f $PYTHON_REQUIREMENTS_FILE ]; then
	REQUIREMENTS_FILE=$PYTHON_REQUIREMENTS_FILE
else
	CKAN_MINOR_VERSION=$(#{new_resource.virtualenv_dir}/bin/python -c "import ckan; print(ckan.__version__)" | grep -o '^[0-9]*[.][0-9]*')
	CKAN_REQUIREMENTS_FILE=requirements-$CKAN_MINOR_VERSION.txt
	if [ -f "$CKAN_REQUIREMENTS_FILE" ]; then
		REQUIREMENTS_FILE=$CKAN_REQUIREMENTS_FILE
	else
		REQUIREMENTS_FILE=requirements.txt
	fi
fi
if [ -f "$REQUIREMENTS_FILE" ]; then
	REQUIREMENTS_FILES="-r $REQUIREMENTS_FILE"
fi
# ckanext-harvest uses this filename
if [ -f "pip-requirements.txt" ]; then
	REQUIREMENTS_FILES="$REQUIREMENTS_FILES -r pip-requirements.txt"
fi
pip install -e . $REQUIREMENTS_FILES || exit 1
```

Reply:
@duttonw I agree installing the dependencies in master it's not ideal. But until dependencies are [packaged as part of the published package](https://github.com/ckan/ckan/issues/8382#issuecomment-2353188970) and installed as part of `pip install ckanext-xloader` is perhaps the least confusing option with the suggested install method. Otherwise I'd suggest the clone, install editable, install requirements.txt approach that we use on other extensions:

```
git clone https://github.com/ckan/ckanext-xloader
cd ckanext-xloader
pip install -e .
pip install -r requirements.txt
```

If I'm not mistaken the script you shared deals with finding the right requirements file locally but not choosing the right one for a particular version of the extension

Reply:
> If I'm not mistaken the script you shared deals with finding the right requirements file locally but not choosing the right one for a particular version of the extension

No, it should handle that. If the requirements are defined correctly, then checking out a specific commit should have requirements that match the code.

The point of the change is to be able to handle multiple external dependencies, eg different CKAN versions and/or Python versions.

Reply:
Exactly, but the current suggested install method (`pip install ckanext-xloader` or `pip install git+https://github.com/ckan/ckanext-xloader@master`) won't check out the code in a place where the user or a script can find it (it will go to the site-packages folder of the venv), so we can not point them to an actual local requirements.txt to run `pip install -r ...`

I'm happy with whatever you think makes more sense, I just don't know how to translate this script into actual instructions.

To be honest, I think xloader would be fine defining its requirements in pyprojec.toml using the [proposed guidelines](https://github.com/ckan/ckan/issues/8382#issuecomment-2353188970) as most of the pinned stuff in requirements.txt are CKAN core requirements anyway

Reply:
yep, Transition to pyproject.toml is on the cards. Am planning on borrowing how https://github.com/DataShades/ckanext-oidc-pkce/commit/c0271584479349d0d8b986b0030c1d637ff3a0d1 this project did it in the near future.